client/asset/dcr/spv.go

```diff
@@ -0,0 +1,1020 @@
+}
+
+// PeerCount returns the count of currently connected peers.
+func (w *spvWallet) PeerCount(ctx context.Context) (uint32, error) {
+	return uint32(len(w.spv.GetRemotePeers())), nil
+}
+
+// cleanBlockCache discards from the blockCache any blocks that have not been
+// accessed for > 2 hours.
+func (w *spvWallet) cleanBlockCache() {
+	w.blockCache.Lock()
+	defer w.blockCache.Unlock()
+	for blockHash, cb := range w.blockCache.blocks {
+		if time.Since(cb.lastAccess) > time.Hour*2 {
+			delete(w.blockCache.blocks, blockHash)
+		}
+	}
+}
+
+func newSpvSyncer(w *wallet.Wallet, netDir string, connectPeers []string) *spv.Syncer {
+	addr := &net.TCPAddr{IP: net.ParseIP("::1"), Port: 0}
+	amgr := addrmgr.New(netDir, net.LookupIP)
+	lp := p2p.NewLocalPeer(w.ChainParams(), addr, amgr)
+	syncer := spv.NewSyncer(w, lp)
+	if len(connectPeers) > 0 {
+		syncer.SetPersistentPeers(connectPeers)
+	}
+	w.SetNetworkBackend(syncer)
+	return syncer
+}
+
+// extendAddresses ensures that the internal and external branches have been
+// extended to the specified indices. This can be used at wallet restoration to
+// ensure that no duplicates are encountered with existing but unused addresses.
+func extendAddresses(ctx context.Context, extIdx, intIdx uint32, dcrw *wallet.Wallet) error {
+	currentExt, currentInt, err := dcrw.BIP0044BranchNextIndexes(ctx, defaultAcct)
+	if err != nil {
+		return err
+	}
+
+	toGo := currentExt - extIdx
+	if toGo > 0 {
+		for i := uint32(0); i < toGo; i++ {
+			_, err := dcrw.NewExternalAddress(ctx, defaultAcct)
+			if err != nil {
+				return fmt.Errorf("NewExternalAddress error: %w", err)
+			}
+		}
+	}
+
+	toGo = currentInt - intIdx
+	if toGo > 0 {
+		for i := uint32(0); i < toGo; i++ {
+			_, err := dcrw.NewInternalAddress(ctx, defaultAcct)
+			if err != nil {
+				return fmt.Errorf("NewInternalAddress error: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func newWalletConfig(db wallet.DB, chainParams *chaincfg.Params) *wallet.Config {
+	return &wallet.Config{
+		DB:              db,
+		GapLimit:        defaultGapLimit,
+		AccountGapLimit: defaultAccountGapLimit,
+		ManualTickets:   defaultManualTickets,
+		AllowHighFees:   defaultAllowHighFees,
+		RelayFee:        defaultRelayFeePerKb,
+		Params:          chainParams,
+	}
+}
+
+func checkCreateDir(path string) error {
+	if fi, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			// Attempt data directory creation
+			if err = os.MkdirAll(path, 0700); err != nil {
+				return fmt.Errorf("cannot create directory: %s", err)
+			}
+		} else {
+			return fmt.Errorf("error checking directory: %s", err)
+		}
+	} else {
+		if !fi.IsDir() {
+			return fmt.Errorf("path '%s' is not a directory", path)
+		}
+	}
+
+	return nil
+}
+
+// walletExists returns whether a file exists at the loader's database path.
+// This may return an error for unexpected I/O failures.
+func walletExists(dbDir string) (bool, error) {
+	exists, err := fileExists(filepath.Join(dbDir, walletDbName))
```

Comment:
You wrote `fileExists` right below so you know it always returns false for non-nil error.  Hence,
`return fileExists(filepath.Join(dbDir, walletDbName))`